### DISCUSSION TITLE
Permit negative P(r) in DeerLab

### DISCUSSION THREAD

**Discussion by schmidtnedlitz** [JeschkeLab/DeerLab#458] [Q&A] [answered]
Dear DeerLab,

I am emailing you to find out on how to remove the restriction on the P(r) to be positive that it can actually also take negative values into account for my T1-edited DEER. I remember in the old DeerAnalysis software I saw a comment that restrained it to be positive, but it seems that I am unable to find it in the current DeerLab version. 

I would appreciate if one could implement a comment that permits negative P(r). I dont try to get negative populations but rather a negative modulation of a population as a function of inversion pulse.

thank you for your help,
Thomas

**Comment by HKaras** [comment, marked as answer]
Hi Thomas, 

Many thanks for you question. I have converted it to a discussion so it remains open for any other interested users. 

Whilst DeerLab was primarliy designed to solve non-negative least squares (nnls) problems it can also handle other cases. The distance distribution `P` is actually just a linear parameter of the model with a lower and upper bound. By default in dipolar model we set its lower bound to be 0, but this can be changed to -inf.

This would be some code that would work well. There is no need to change anything in the `fit` function.
`model.P.lb = np.ones(model.P.lb.shape) * -1* np.inf`

Does your problem also permit `P` to be positive or is this an entierly negative `P`?
There does appear to be a bug in the linearly constrained but not non-negative solver, so if you need an upper bound it will not currently work properly. I will try and fix this next week.

**Reply by schmidtnedlitz** [reply]
Thank you Hugo for the response. I will give that a trial.

actually ... when I define lb as -inf I get the following comment: 
""" ValueError: Residuals are not finite in the initial point. """

Well actually P can be both negative and positive in the same trace, therefore let me know when you are able to get your hands on it .... I am looking forward in trying it.

Thank you for your help and contribution,
Thomas

**Reply by HKaras** [reply]
Hi Thomas, 

Sorry for the slow response. I didn't notice that you had edited your message
I haven't been able to reproduce on my computer. Please could you share your code and data. 
If you feel more comfortable doing this in private by email, then feel free to email me at hkaras@ethz.ch

Hugo
